### PR TITLE
Test helpers improvements

### PR DESCRIPTION
**Pull request by jjbustamante**
### References

- This Pull Request is created after [PR #125](https://github.com/buildpacks/imgutil/pull/125) was merged

### Context

The idea behind [PR #125](https://github.com/buildpacks/imgutil/pull/125) was to provide a structure to help us on testing the new validations in the lifecycle described in issue [#572](https://github.com/buildpacks/lifecycle/issues/572). At that point, some improvements were raised and the idea was to simply our docker registry implementation. This pull request is addressing those improvements.

### Description

- The PR exposes a new method `Add` to allow dynamically configure the `ImagePrivileges` in the registry, basically providing an option to include new mappings after the registry was created.
- The PR is also addressing one little problem that avoid us from using only one docker registry implementation to do our tests.

### Actual

Right now, if we need to simulate a read-only registry or a registry based on basic authentication, we need to create two instances similar  to the following image

test-helpers-improvements-1

Basically, if the registry is read-only no authentication check is made. After merging the [PR #125](https://github.com/buildpacks/imgutil/pull/125), we added a third behavior

wrapping
 
As we can see, the problem that avoid us from using the custom registry to implement both behaviors (read-only and basic authentication) is that for the case of read-only we are actually running credential validation!

### After

The PR is implementing the following behavior 

registry-3

We are improving the Custom logic to handle the scenarios:
 * The image is readable
 * The image is writable 
 * Or both
 In this way, we are going to be able to simplify the tests in the lifecycle by removing the docker registry instances and using only one to handle all the test cases

### CODE REVIEW

File: testhelpers/docker_registry.go [hunk 1]
```diff
@@ -68,17 +68,24 @@ func NewDockerRegistry(ops ...RegistryOption) *DockerRegistry {
 }
 
 // BasicAuth wraps a handler, allowing requests with matching username and password headers, otherwise rejecting with a 401
-func BasicAuth(handler http.Handler, username, password, realm string) http.Handler {
+// optPermissions is used to skip the authentication check if the image is readable.
```
Comment (natalieparellano) [inline testhelpers/docker_registry.go:71]:
```suggestion
// optPermissions is used to skip the authentication check if the image has been configured in the test setup to be always readable.
```

File: testhelpers/docker_registry.go [hunk 2]
```diff
@@ -119,6 +126,11 @@ func Custom(handler http.Handler, permissions map[string]ImagePrivileges) http.H
 	})
 }
 
+// Start creates a docker registry following these rules:
+// - Shared handler will be use, otherwise a new one will be created
```
Comment (natalieparellano) [inline testhelpers/docker_registry.go:130]:
```suggestion
// - Shared handler will be used, otherwise a new one will be created
```

File: testhelpers/docker_registry.go [hunk 3]
```diff
@@ -119,6 +126,11 @@ func Custom(handler http.Handler, permissions map[string]ImagePrivileges) http.H
...
+// - by default the shared handler will be wrapped using a read only handler
```
Comment (natalieparellano) [inline testhelpers/docker_registry.go:131]:
```suggestion
// - By default the shared handler will be wrapped using a read only handler
```

File: testhelpers/docker_registry.go [hunk 4]
```diff
@@ -119,6 +126,11 @@ func Custom(handler http.Handler, permissions map[string]ImagePrivileges) http.H
...
+// - In case credentials are configured, the shared handler will be wrapped with a basic authentication handler and
+//   in any image privileges were set, then custom handler will be use to wrap the auth handler.
```
Comment (natalieparellano) [inline testhelpers/docker_registry.go:133]:
```suggestion
//   if any image privileges were set, then the custom handler will be used to wrap the auth handler.
```

File: testhelpers/docker_registry.go [hunk 5]
```diff
@@ -68,17 +68,24 @@ func NewDockerRegistry(ops ...RegistryOption) *DockerRegistry {
 }
 
 // BasicAuth wraps a handler, allowing requests with matching username and password headers, otherwise rejecting with a 401
-func BasicAuth(handler http.Handler, username, password, realm string) http.Handler {
+// optPermissions is used to skip the authentication check if the image is readable.
+func BasicAuth(handler http.Handler, username, password, realm string, optPermissions ...map[string]ImagePrivileges) http.Handler {
```
Comment (natalieparellano) [inline testhelpers/docker_registry.go:72]: Why does `optPermissions` need to be passed in if `customPrivileges` is a field on the registry?

File: testhelpers/image_privileges.go [hunk 6]
```diff
@@ -12,7 +12,7 @@ type ImagePrivileges struct {
 	writable bool
 }
 
-// Creates a new ImagePrivileges, use "readable" or "writable" to set the properties accordingly.
+// NewImagePrivileges Creates a new ImagePrivileges, use "readable" or "writable" to set the properties accordingly.
```
Comment (natalieparellano) [inline testhelpers/image_privileges.go:15]:
```suggestion
// NewImagePrivileges creates a new ImagePrivileges, use "readable" or "writable" to set the properties accordingly.
```

File: testhelpers/image_privileges.go [hunk 7]
```diff
@@ -24,11 +24,11 @@ func NewImagePrivileges(imageName string) (priv ImagePrivileges) {
 	return
 }
 
-// Based of the registry API specification https://docs.docker.com/registry/spec/api/
-// This method returns the image name for path value that matches requests to blobs, manifests or tags
+// extractImageName returns the image name for path value that matches requests to blobs, manifests or tags
```
Comment (natalieparellano) [inline testhelpers/image_privileges.go:27]:
```suggestion
// extractImageName returns the image name from a path value that matches requests to blobs, manifests or tags
```

File: testhelpers/image_privileges.go [hunk 8]
```diff
@@ -24,11 +24,11 @@ func NewImagePrivileges(imageName string) (priv ImagePrivileges) {
...
 // For examples:
 // extractImageName("v2/foo.bar/blobs/") returns "foo.bar"
 // extractImageName("v2/foo/bar/manifests/") returns "foo/bar"
+// Based of the registry API specification https://docs.docker.com/registry/spec/api/
```
Comment (natalieparellano) [inline testhelpers/image_privileges.go:31]:
```suggestion
// Based on the Docker registry API specification: https://docs.docker.com/registry/spec/api/
```

**Review by natalieparellano — COMMENTED** [review]
@jjbustamante this will be a great improvement :)

Added a couple questions & nits.

File: testhelpers/docker_registry.go (hunk 5 above)
Comment (jjbustamante) [reply testhelpers/docker_registry.go:72]: Because these HTTP handler functions don't have access to those attributes, they are plain functions handling the HTTP request. I think that's why we are also sending the username/password values from the registry.

Comment (natalieparellano) [reply testhelpers/docker_registry.go:72]: Ah so again, reading more closely, it would be called before `Start()`.

File: testhelpers/docker_registry.go [hunk 9]
```diff
@@ -182,6 +198,13 @@ func (r *DockerRegistry) EncodedLabeledAuth() string {
 	return base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf(`{"username":"%s","password":"%s"}`, r.username, r.password)))
 }
 
+// Add stores the given key name with the provided ImagePrivileges. For example
+// Add("my-image", NewImagePrivileges("foo-writable-readable")) will save "my-image" as a
+// readable and writable image into the registry
```
Comment (natalieparellano) [inline testhelpers/docker_registry.go:203]:
```suggestion
// readable and writable image into the registry.
// It must be called before Start to have any effect.
```

File: testhelpers/docker_registry.go (hunk 5 above)
Comment (jjbustamante) [reply testhelpers/docker_registry.go:72]: Actually, if I am not wrong, the HTTP server is holding the reference to the handler here:
```	
r.server = &httptest.Server{
	Listener: listener,
	Config:   &http.Server{Handler: r.authnHandler},
}
```
every time a new request is coming, the HTTP server is invoking the handler function which is, in fact, referencing the attribute from the docker registry (the permission map), and that is how it is dynamically updated when call the `Add` method  after the `Start`

File: testhelpers/image_privileges.go (hunk 6 above)
Comment (natalieparellano) [reply testhelpers/image_privileges.go:15]: If I'm understanding https://github.com/buildpacks/lifecycle/pull/685/files correctly, we don't actually care that `imageName` is an image name. We just care what permissions we want the image to have when we call `testRegistry.Add(myImage, ih.NewImagePrivileges(readWriteImage)))`.

If that's true why not make `readable` and `writable` constants rather than relying on this fuzzy matching? (Here's a playground showing what I'm thinking: https://play.golang.org/p/tR0h1HFW8P- )

File: testhelpers/docker_registry.go (hunk 1 above)
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: Fixed in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/1a4373ac85d225b6bd3771c3ccadae2a239cf6db)

File: testhelpers/docker_registry.go (hunk 2 above)
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: Fixed in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/1a4373ac85d225b6bd3771c3ccadae2a239cf6db)

File: testhelpers/docker_registry.go (hunk 3 above)
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: Fixed in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/1a4373ac85d225b6bd3771c3ccadae2a239cf6db)

File: testhelpers/docker_registry.go (hunk 4 above)
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: Fixed in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/1a4373ac85d225b6bd3771c3ccadae2a239cf6db)

File: testhelpers/image_privileges.go (hunk 7 above)
Comment (jjbustamante) [reply testhelpers/image_privileges.go]: Fixed in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/1a4373ac85d225b6bd3771c3ccadae2a239cf6db)

File: testhelpers/image_privileges.go (hunk 8 above)
Comment (jjbustamante) [reply testhelpers/image_privileges.go]: Fixed in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/1a4373ac85d225b6bd3771c3ccadae2a239cf6db)

File: testhelpers/docker_registry.go (hunk 9 above)
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: Fixed in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/1a4373ac85d225b6bd3771c3ccadae2a239cf6db)

File: testhelpers/image_privileges.go (hunk 6 above)
Comment (jjbustamante) [reply testhelpers/image_privileges.go]: Yeah, I like it!

Comment (jjbustamante) [reply testhelpers/image_privileges.go]: I added those changes in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/1924767685e509286c1db17a8e2c2f791113a114)

File: testhelpers/image_privileges.go [hunk 10]
```diff
@@ -12,23 +12,36 @@ type ImagePrivileges struct {
...
-func NewImagePrivileges(imageName string) (priv ImagePrivileges) {
-	priv.readable = strings.Contains(imageName, "readable")
-	priv.writable = strings.Contains(imageName, "writable")
+// NewImagePrivileges() returns ImagePrivileges{readable: false, writable: false}
+// NewImagePrivileges(Readable) returns ImagePrivileges{readable: true, writable: false}
+// NewImagePrivileges(Writable) returns ImagePrivileges{readable: false, writable: true}
+// NewImagePrivileges(Readable, Writable) returns ImagePrivileges{readable: true, writable: true}
+func NewImagePrivileges(imageAccess ...ImageAccess) (priv ImagePrivileges) {
+	for _, ia := range imageAccess {
+		switch ia {
```
Comment (yaelharel) [inline testhelpers/image_privileges.go:30]: I think you should add a `default` case with a warning message, since theoretically someone can pass a value that isn't `Readable` nor `Writable` (because `ImageAccess` is an `int`).

File: testhelpers/docker_registry.go [hunk 11]
```diff
@@ -182,6 +199,14 @@ func (r *DockerRegistry) EncodedLabeledAuth() string {
 	return base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf(`{"username":"%s","password":"%s"}`, r.username, r.password)))
 }
 
+// Add stores the given key name with the provided ImagePrivileges. For example
+// Add("my-image", NewImagePrivileges(Readable, Writable)) will save "my-image" as a
+// readable and writable image into the registry.
+// It must be called before Start to have any effect.
```
Comment (yaelharel) [inline testhelpers/docker_registry.go:205]: I don't think it's a good practice to ask the user to remember to call `Add` before calling `Start`.
It might be very confusing to find out the reason that your code isn't running the way you expect it to be because of such thing.
Maybe it's better to pass the `ImagePrivileges` to the `Start` function, or have two `Start` functions:
`StartWithPrivileges` and `StartWithoutPrivileges`.
These are just two ideas. Maybe there are better ones.

By the way, maybe I didn't follow but it looks like you're calling the `Add` function after the `Start` function in your [lifecycle PR](https://github.com/buildpacks/lifecycle/pull/685/files).

**Review by yaelharel — COMMENTED** [review]
Hi @jjbustamante, these are great improvements.
Please see my comments.
Moreover, there are a few open discussions from your previous PR that are still relevant and I think we should consider continue discussing here (or open new issues for them) - https://github.com/buildpacks/imgutil/pull/125/files.
Thanks!

File: testhelpers/docker_registry.go (hunk 5 above)
Comment (jjbustamante) [reply testhelpers/docker_registry.go:73]: @yaelharel , @natalieparellano.

The `Add` method can actually be called after the `Start`, as I mentioned before, it handler is holding a reference to the internal Map and every time the internal map is being updated after an `Add` invocation when the handler function is called the Map update is reflected.

I added a log entry in the Custom handler function to print the URL of the request and the content of the Map that is received, as you can see, the Map values are updated and the handler is called on every request made to the Docker registry on each test cases.

[local-out.txt](https://github.com/buildpacks/imgutil/files/6970269/local-out.txt)

File: testhelpers/docker_registry.go [hunk 12]
```diff
@@ -68,17 +68,25 @@ func NewDockerRegistry(ops ...RegistryOption) *DockerRegistry {
...
+func BasicAuth(handler http.Handler, username, password, realm string, optPermissions ...map[string]ImagePrivileges) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if len(optPermissions) > 0 && (r.Method == "GET" || r.Method == "HEAD") {
+			permissions := optPermissions[0]
+			if permission, ok := permissions[extractImageName(r.URL.Path)]; ok {
+				if permission.readable {
+					handler.ServeHTTP(w, r)
+				}
+			}
+		}
```
Comment (natalieparellano) [inline testhelpers/docker_registry.go:82]: This sort of bothers me in that we're duplicating logic in `Custom()` (and also not handling the case when we want to "pass" a writeable image). It seems that what we want to do is IF the url is found in the map, just fall through to the regular handler (if permission is not denied). What if instead, we send `r.regHandler` into `Custom()` - and then around line 124, instead of falling through to the auth handler, we invoke `regHandler.Serve()`. Then we could delete this logic.

File: testhelpers/docker_registry.go (hunk 11 above)
Comment (natalieparellano) [reply testhelpers/docker_registry.go:205]: Whoops, sorry, that was my code suggestion when I didn't understand how it worked!

**Comment by natalieparellano** [comment]
@jjbustamante this is looking pretty good to me - LMK what you think of https://github.com/buildpacks/imgutil/pull/126/files#r687051113

File: testhelpers/image_privileges.go [hunk 13]
```diff
@@ -25,16 +26,19 @@ const (
 // NewImagePrivileges(Readable) returns ImagePrivileges{readable: true, writable: false}
 // NewImagePrivileges(Writable) returns ImagePrivileges{readable: false, writable: true}
 // NewImagePrivileges(Readable, Writable) returns ImagePrivileges{readable: true, writable: true}
-func NewImagePrivileges(imageAccess ...ImageAccess) (priv ImagePrivileges) {
+func NewImagePrivileges(imageAccess ...ImageAccess) (ImagePrivileges) {
```
Comment (yaelharel) [inline testhelpers/image_privileges.go:29]: You don't need the `()` around the return type

File: testhelpers/docker_registry.go (hunk 12 above)
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: Keeping my idea of expressing the behavior with a diagram, doing this will work like this:

registry-3

The Custom handler decides which one to call based on the privileges and skipping the Basic Auth if it is a read operation. Yeap, I like it

File: testhelpers/docker_registry.go (hunk 11 above)
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: I deleted the comment in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/1c2361a45515bb01cc7fe1fae051d620fdc643f3)

File: testhelpers/docker_registry.go [hunk 14]
```diff
@@ -96,11 +94,13 @@ func ReadOnly(handler http.Handler) http.Handler {
...
 }
 
-func Custom(handler http.Handler, permissions map[string]ImagePrivileges) http.Handler {
+func Custom(basicAuthHandler http.Handler, regHandler http.Handler, permissions map[string]ImagePrivileges) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		if permission, ok := permissions[extractImageName(r.URL.Path)]; ok {
 			if r.Method == "GET" || r.Method == "HEAD" {
-				if !permission.readable {
+				if permission.readable {
+					regHandler.ServeHTTP(w, r)
```
Comment (natalieparellano) [inline testhelpers/docker_registry.go:102]: Wouldn't we also want to "wave through" a put request where the image has been explicitly configured to be writeable? Should we require basic auth in that case (I'd argue no). So there would be a similar extra lines right around line 109...

File: testhelpers/docker_registry.go [hunk 15]
```diff
@@ -135,10 +140,14 @@ func (r *DockerRegistry) Start(t *testing.T) {
 	r.authnHandler = ReadOnly(r.regHandler)
 	if r.username != "" {
 		if r.customPrivileges != nil {
+			// wrap registry handler with basic auth
+			r.authnHandler = BasicAuth(r.regHandler, r.username, r.password, "registry")
 			// wrap registry handler with custom handler
-			r.regHandler = Custom(r.regHandler, r.customPrivileges)
+			r.authnHandler = Custom(r.authnHandler, r.regHandler, r.customPrivileges)
```
Comment (natalieparellano) [inline testhelpers/docker_registry.go:146]:
```suggestion
			r.authnHandler = Custom(basicAuthHandler, r.regHandler, r.customPrivileges)
```

Comment (natalieparellano) [reply testhelpers/docker_registry.go:146]: (I think this makes it a little more clear)

File: testhelpers/docker_registry.go (hunk 14 above)
Comment (jjbustamante) [reply testhelpers/docker_registry.go:102]: If the image is writable, it will delegate the request to the Basic Auth handler that will check if you sent the credentials in the request, like the following image:

registry-3

Comment (jjbustamante) [reply testhelpers/docker_registry.go:102]: The idea of all this stuff, was because to implement the read-only behavior without forcing the client to send the Basic Auth value in the request, but the Write operation must validate the Auth, right? or you mean if the image is writable (by anyone) we can skip the validation? Not sure if we can do that in the registry right now, in that case, yes! we can add a similar logic in line 109

Comment (natalieparellano) [reply testhelpers/docker_registry.go:102]: That's interesting - I think you could make the case for it working either way. It's probably worth a quick discussion!

File: testhelpers/image_privileges.go (hunk 10 above)
Comment (jjbustamante) [reply testhelpers/image_privileges.go:32]: Fixed in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/9fcb86825b0095e4dc69b5c128a8400944d50f04)

File: testhelpers/image_privileges.go [hunk 16]
```diff
@@ -12,23 +12,36 @@ type ImagePrivileges struct {
...
-// NewImagePrivileges("foo-writable") returns ImagePrivileges{readable: false, writable: true}
-// NewImagePrivileges("foo-writable-readable") returns ImagePrivileges{readable: true, writable: true}
-func NewImagePrivileges(imageName string) (priv ImagePrivileges) {
-	priv.readable = strings.Contains(imageName, "readable")
-	priv.writable = strings.Contains(imageName, "writable")
+// NewImagePrivileges() returns ImagePrivileges{readable: false, writable: false}
+// NewImagePrivileges(Readable) returns ImagePrivileges{readable: true, writable: false}
+// NewImagePrivileges(Writable) returns ImagePrivileges{readable: false, writable: true}
+// NewImagePrivileges(Readable, Writable) returns ImagePrivileges{readable: true, writable: true}
+func NewImagePrivileges(imageAccess ...ImageAccess) (priv ImagePrivileges) {
```
[the remark this answers is not in the document]
Comment (jjbustamante) [reply testhelpers/image_privileges.go]: Fixed in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/1c2361a45515bb01cc7fe1fae051d620fdc643f3)

File: testhelpers/image_privileges.go (hunk 13 above)
Comment (jjbustamante) [reply testhelpers/image_privileges.go]: Removed

File: testhelpers/docker_registry.go [hunk 17]
```diff
@@ -135,10 +140,14 @@ func (r *DockerRegistry) Start(t *testing.T) {
 	r.authnHandler = ReadOnly(r.regHandler)
 	if r.username != "" {
 		if r.customPrivileges != nil {
+			// wrap registry handler with basic auth
+			r.authnHandler = BasicAuth(r.regHandler, r.username, r.password, "registry")
```
[the remark this answers is not in the document]
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: I agreed, applied in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/f19016e4dd378dcd0531ef1d351626730dd1ccb7)

File: testhelpers/docker_registry.go (hunk 15 above)
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: I agreed, applied in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/f19016e4dd378dcd0531ef1d351626730dd1ccb7)

File: testhelpers/docker_registry.go (hunk 14 above)
Comment (natalieparellano) [reply testhelpers/docker_registry.go:102]: I think we might need a `return` after we call `regHandler.ServeHTTP()`. What do you think of the following?

```
var (
			image ImagePrivileges
			ok    bool
		)
		if image, ok = permissions[extractImageName(r.URL.Path)]; !ok {
			basicAuthHandler.ServeHTTP(w, r)
			return
		}
		if r.Method == "GET" || r.Method == "HEAD" {
			if !image.readable {
				w.WriteHeader(401)
				_, _ = w.Write([]byte("Unauthorized.\n"))
				return
			}
		} else { // assume write request
			if !image.writable {
				w.WriteHeader(401)
				_, _ = w.Write([]byte("Unauthorized.\n"))
				return
			}
		}
		regHandler.ServeHTTP(w, r)
```

File: testhelpers/docker_registry.go [hunk 18]
```diff
@@ -193,11 +192,37 @@ func (r *DockerRegistry) EncodedLabeledAuth() string {
 	return base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf(`{"username":"%s","password":"%s"}`, r.username, r.password)))
 }
 
-// Add stores the given key name with the provided ImagePrivileges. For example
-// Add("my-image", NewImagePrivileges(Readable, Writable)) will save "my-image" as a
+// SetPrivilegesToImage sets the given image name with the provided ImagePrivileges if ImagePrivileges was enabled.
```
Comment (yaelharel) [inline testhelpers/docker_registry.go:196]: What do you mean by "if ImagePrivileges was enabled"?

File: testhelpers/docker_registry.go [hunk 19]
```diff
@@ -182,6 +193,39 @@ func (r *DockerRegistry) EncodedLabeledAuth() string {
...
+// For example SetPrivilegesToImage("my-image", NewImagePrivileges(Readable, Writable)) will save "my-image" as a
+// readable and writable image into the registry.
+func (r *DockerRegistry) setPrivilegesToImage(imageName string, privilege ImagePrivileges) {
+	if r.isImagePrivilegesEnable() {
+		r.imagePrivileges[imageName] = privilege
+	}
+}
+
+// MakeReadOnly set the given image name to be readable when the ImagePrivileges feature was enabled
+func (r *DockerRegistry) MakeReadOnly(imageName string) {
```
Comment (yaelharel) [inline testhelpers/docker_registry.go:206]:
```suggestion
func (r *DockerRegistry) SetReadOnly(imageName string) {
```

Comment (yaelharel) [reply testhelpers/docker_registry.go:206]: If you agree with this suggestion, I would change the other functions as well.

File: testhelpers/docker_registry.go [hunk 20]
```diff
@@ -96,29 +95,38 @@ func ReadOnly(handler http.Handler) http.Handler {
 	})
 }
 
-func Custom(handler http.Handler, permissions map[string]ImagePrivileges) http.Handler {
+func delegator(basicAuthHandler http.Handler, regHandler http.Handler, permissions map[string]ImagePrivileges) http.Handler {
```
Comment (yaelharel) [inline testhelpers/docker_registry.go:98]: I like the refactor in this function!

File: testhelpers/docker_registry.go [hunk 21]
```diff
@@ -182,6 +193,39 @@ func (r *DockerRegistry) EncodedLabeledAuth() string {
...
+func (r *DockerRegistry) MakeReadWrite(imageName string) {
+	r.setPrivilegesToImage(imageName, NewImagePrivileges(Readable, Writable))
+}
+
+// MakeInaccessible set the given image name to do not have any access when the ImagePrivileges feature was enabled
+func (r *DockerRegistry) MakeInaccessible(imageName string) {
+	r.setPrivilegesToImage(imageName, NewImagePrivileges())
+}
+
+func (r *DockerRegistry) isImagePrivilegesEnable() bool {
```
Comment (yaelharel) [inline testhelpers/docker_registry.go:225]: I think that it should be: `isImagePrivilegesEnabled`

File: testhelpers/docker_registry.go [hunk 22]
```diff
@@ -96,29 +95,38 @@ func ReadOnly(handler http.Handler) http.Handler {
 	})
 }
 
-func Custom(handler http.Handler, permissions map[string]ImagePrivileges) http.Handler {
+func delegator(basicAuthHandler http.Handler, regHandler http.Handler, permissions map[string]ImagePrivileges) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (yaelharel) [inline testhelpers/docker_registry.go:99]: I think that I would change `r` and `w` to `req` and `res` or even to `request` and `response` (or maybe you have better names?).
While working on "read/write" permissions, it's a bit confusing that `w` doesn't influence the permissions because it's just the response writer.

File: testhelpers/docker_registry.go (hunk 21 above)
Comment (natalieparellano) [reply testhelpers/docker_registry.go:225]: I like `imagePrivilegesEnabled`

File: testhelpers/docker_registry.go [hunk 23]
```diff
@@ -182,6 +193,39 @@ func (r *DockerRegistry) EncodedLabeledAuth() string {
...
+// readable and writable image into the registry.
+func (r *DockerRegistry) setPrivilegesToImage(imageName string, privilege ImagePrivileges) {
+	if r.isImagePrivilegesEnable() {
+		r.imagePrivileges[imageName] = privilege
+	}
+}
+
+// MakeReadOnly set the given image name to be readable when the ImagePrivileges feature was enabled
+func (r *DockerRegistry) MakeReadOnly(imageName string) {
+	r.setPrivilegesToImage(imageName, NewImagePrivileges(Readable))
```
Comment (natalieparellano) [inline testhelpers/docker_registry.go:207]:
```suggestion
	r.setImagePrivileges(imageName, NewImagePrivileges(Readable))
```
(nit)

File: testhelpers/docker_registry.go (hunk 19 above)
Comment (natalieparellano) [reply testhelpers/docker_registry.go:206]: This is pure syntactic sugar, but maybe you could make this function return the "repo name" for the image, so that [these lines](https://github.com/buildpacks/lifecycle/pull/685/files#diff-c4e4b4a55ebbd71443c5260020b287f3018cb7bb4c65c79e74554b7f97c45656R168-R169) could become 
```
fixtures.readOnlyRegCacheImage = testRegistry.MakeReadWrite(someCacheName) // or SetReadWrite

**Review by natalieparellano — APPROVED** [review]
@jjbustamante this is great! I agree with many of @yaelharel's suggestions for making the code more readable, but I also think we could merge it as-is (let us know what you'd like to do). It will be awesome to put these helpers to use in the lifecycle, this is going to simplify our tests so much! Thank you!

File: testhelpers/docker_registry.go (hunk 19 above)
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: I like making the method to return the repoName. I added in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/c59c94c0966e2826f3e1ce6b48983be810952fee)

File: testhelpers/docker_registry.go [hunk 24]
```diff
@@ -96,29 +95,38 @@ func ReadOnly(handler http.Handler) http.Handler {
...
-				}
+		var (
+			image ImagePrivileges
+			ok    bool
+		)
+		if image, ok = permissions[extractImageName(r.URL.Path)]; !ok {
+			basicAuthHandler.ServeHTTP(w, r)
+			return
+		}
+		if r.Method == "GET" || r.Method == "HEAD" {
```
[the remark this answers is not in the document]
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: Yes, it is also fine to add that method. I added in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/c59c94c0966e2826f3e1ce6b48983be810952fee)

File: testhelpers/docker_registry.go (hunk 21 above)
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: Renamed to `imagePrivilegesEnabled` in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/c59c94c0966e2826f3e1ce6b48983be810952fee)

File: testhelpers/docker_registry.go (hunk 22 above)
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: Cool, I renamed them in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/c59c94c0966e2826f3e1ce6b48983be810952fee)

File: testhelpers/docker_registry.go (hunk 23 above)
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: Fixed in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/c59c94c0966e2826f3e1ce6b48983be810952fee)

File: testhelpers/docker_registry.go (hunk 18 above)
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: Well, because the method is only exposed in the package scope I added `if WithImagePrivileges was called during registry creation`. Is it clear?

File: testhelpers/docker_registry.go [hunk 25]
```diff
@@ -71,54 +72,61 @@ func NewDockerRegistry(ops ...RegistryOption) *DockerRegistry {
...
-
 		handler.ServeHTTP(w, r)
 	})
 }
 
 // ReadOnly wraps a handler, allowing only GET and HEAD requests, otherwise rejecting with a 405
 func ReadOnly(handler http.Handler) http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		if !(r.Method == "GET" || r.Method == "HEAD") {
+	return http.HandlerFunc(func(w http.ResponseWriter, request *http.Request) {
```
Comment (yaelharel) [inline testhelpers/docker_registry.go:87]:
```suggestion
	return http.HandlerFunc(func(response http.ResponseWriter, request *http.Request) {
```

**Review by yaelharel — APPROVED** [review]
Added a small suggestion but other than that, your PR looks great!
Thank you 😊

File: testhelpers/docker_registry.go (hunk 25 above)
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: Oh, I missed this one. Sorry it's fixed in the following [commit](https://github.com/buildpacks/imgutil/pull/126/commits/1eb46031019fbbfccbc8f307cb6d467508e41a7d)
